### PR TITLE
refactor+style+workflowbuild+versionbump: Remove needless code nesting

### PR DESCRIPTION
**Pull request by HugoVG**
## In this PR:
- Bumped to dotnet 7 to take advantage of all that dotnet 7 goodness.
- made a workflow file to test build(more coming to test)
- Changed the Main MD file to be a little more tech illiterate friendly
- Adding option for Wiggle when talking(works but i need to make it a little better still)
- some code changes that remove some really nested code 
- Discord addin
- Multiple "Visemes"(pictures) inside the CharacterAnimatorComponent
- solved a Small issue where the DragFloat would be immensly streched out inside a 4k monitor

More PR's will be coming, but i'll merge this for the time being.

! keep in mind this version HAS the Discord plugin added in the components, when i get a plugin loader to work i'm move this to a seperate area (also why there is a README.md in the folder)

### CODE REVIEW

**Comment by HugoVG** [comment]
Do you want [Avalonia](https://avaloniaui.net/) as framework for it?

**Comment by naomiEve** [comment]
> Do you want [Avalonia](https://avaloniaui.net/) as framework for it?

So far we're using the bindings for DearImGui which is fairly small and platform-agnostic, as it only requires a drawable surface. My initial plan was to style ImGui to look presentable (kind of like how InochiCreator does it), but if other UI toolkits are an option, I'd be interested too.

The only thing I'd really want, though, is to keep the application fairly small. One of the goals in VTubeTiny is ease of installation and portability, and keeping the file size as small as it can be is one of the biggest factors in that.

**Review by naomiEve — CHANGES_REQUESTED** [review]
Before merging, please rename these commits to adhere to the Conventional Commits standard, as described here: [https://www.conventionalcommits.org/en/v1.0.0/](https://www.conventionalcommits.org/en/v1.0.0/).

So for example, the title of this commit would be something like: `refactor+style: Remove needless code nesting.` :)

File: VTTiny/Editor/EditorGUI.cs
Comment (naomiEve) [inline VTTiny/Editor/EditorGUI.cs:265]: For if, or for(each) statements whose body only takes up one line, please remove the parentheses before and after and keep it to only one indented line. Thanks!

File: VTTiny/Components/Animator/AnimatorCharacter.Serialization.cs
Comment (naomiEve) [inline VTTiny/Components/Animator/AnimatorCharacter.Serialization.cs:16]: You can safely skip the variable type on declarations like this, and just substitute it with `var`

File: VTTiny/Components/Animator/IPackageAble.cs
Comment (naomiEve) [inline VTTiny/Components/Animator/IPackageAble.cs:5]: I think the proper capitalization here should be `IPackageable`, in order to fit with everything else.

File: VTTiny/Components/Animator/AnimatorCharacter.Editor.cs
Comment (HugoVG) [reply VTTiny/Components/Animator/AnimatorCharacter.Editor.cs:82]: oh no, this is not for adding a new texture, this will make a new place holder to put a new texture in, this is so you can have 9999 different kinds of mouth movements, this is just there in case there is 0 in the list, it'll make a temporary one

**Comment by HugoVG** [comment]
> So far we're using the bindings for DearImGui which is fairly small and platform-agnostic, as it only requires a drawable surface. My initial plan was to style ImGui to look presentable (kind of like how InochiCreator does it), but if other UI toolkits are an option, I'd be interested too.


> The only thing I'd really want, though, is to keep the application fairly small. One of the goals in VTubeTiny is ease of installation and portability, and keeping the file size as small as it can be is one of the biggest factors in that.
Avalonia should be pretty small as well since it works on embeded systems
image
>  One of the goals in VTubeTiny is ease of installation and portability
i'm 100% with you on that part, i want it as straight forward as possible so you don't have to puzzle around trying to get the optimal pngtube layout

**Comment by naomiEve** [comment]
> > The only thing I'd really want, though, is to keep the application fairly small. One of the goals in VTubeTiny is ease of installation and portability, and keeping the file size as small as it can be is one of the biggest factors in that.
> > Avalonia should be pretty small as well since it works on embeded systems
> 
> image
> 
> One of the goals in VTubeTiny is ease of installation and portability i'm 100% with you on that part, i want it as straight forward as possible so you don't have to puzzle around trying to get the optimal pngtube layout

Keep in mind that the entire rendering is done by Raylib (both for the editor and for the stage). If there is some way we can combine Avalonia for rendering the editor's GUI, while also mixing in the Raylib rendering of the stage, it'd be an interesting path forward for the editor.

**Comment by HugoVG** [comment]
i don't really see a quick thing that allows it to run *inside* of Avalonia, could always have a seperate window for that(not as clean) but i'll look into getting it inside of it

**Comment by naomiEve** [comment]
Hmm, in this case I'd like to maybe keep a potential UI rework in a separate PR to this one. I'm very much open to making the UI look better, since the current one is very barren and unfriendly to users, but I don't really want a UI rework to also require us to switch the rendering engine.

**Comment by HugoVG** [comment]
Is there a central place for plugins, or do i put it inside this repo?
currently working on a Discord plugin so friends inside the vc can have their own character that can talk/move

**Comment by naomiEve** [comment]
> Is there a central place for plugins, or do i put it inside this repo? currently working on a Discord plugin so friends inside the vc can have their own character that can talk/move

There's no central place for plugins yet, but this plugin sounds helpful enough that you can for now put it as a separate project within the VTubeTiny solution. :)

**Comment by HugoVG** [comment]
Do you have any objections of me making a github workflow that Build/Publishes the solution?

**Comment by naomiEve** [comment]
Go right ahead! I didn't set up one yet, as I have no idea how they work like.

**Comment by HugoVG** [comment]
Just came to a realization how massively have plugins get due to needing references to ``Vtubetiny`` project, seperating them is whole different deal,
for plugins to work, ``component`` is made inside the main code base so all plugins have to reference it, making it make use all dll's the main project uses, which it then tries to load

**Comment by naomiEve** [comment]
Hmm, honestly I've never tested plugins beyond a couple of small tests, and even then I basically just referenced the compiled VTubeTiny assembly dll. Might be a good idea to look into a better way of exposing the plugin interface.

File: VTTiny.Plugin.Discord/Services/Token.cs
Comment (naomiEve) [inline VTTiny.Plugin.Discord/Services/Token.cs:45]: I know that this is a WIP, but please don't be rude to the user :). A more suitable name would be `TokenNotFoundException`.

Comment (HugoVG) [reply VTTiny.Plugin.Discord/Services/Token.cs:45]: oh i forgot, i copy pasted most code from a previous bot i've made

Comment (BeornOG) [reply VTTiny.Plugin.Discord/Services/Token.cs:45]: 10/10 reusing an originally only for internal usage error

**Comment by HugoVG** [comment]
Because a side effect of linking the main project is that it'll produce a ton of dll's, i think we could look for dll named ``Plugin\.VTTiny.*.dll`` so it won't attempt to load let's say ``Microsoft.Linq``

**Review by naomiEve — CHANGES_REQUESTED** [review]
Before anything, please rename the commits themselves as well to fit the conventional commits standard, and not only the title of the pull request :) Sorry for being overly pedantic about this, I don't wanna mix different naming conventions.

I'll do a proper review of the code tomorrow, thank you once again for your work!

[https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/changing-a-commit-message](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/changing-a-commit-message)

**Comment by HugoVG** [comment]
the 8 commits are kinda to much to rename, i'll probbably just squash all changes into the PR name after the code review

**Review by naomiEve — CHANGES_REQUESTED** [review]
Initial review :) Mostly things related to style, couldn't test the discord audio functionality yet.

Also, please do get clean up the compiler warnings, the discord plugin code generates a loooot of these. Thank you!

**Comment by naomiEve** [comment]
> the 8 commits are kinda to much to rename, i'll probbably just squash all changes into the PR name after the code review

Sure, this works.

**Comment by BeornOG** [comment]
> Mostly things related to style, couldn't test the discord audio functionality yet

i was with hogo in a vc while testing, and it definetly did work on his machine.

**Comment by HugoVG** [comment]
@naomiEve can you commit your .editorconfig file with your code style, I'll just make a runner that auto formats all this on pull request.

Also, the discord stuff will always give "unused code" since they are ``IHandlers``. will remove the warnings.

File: VTTiny/Components/Component.cs
Comment (HugoVG) [reply VTTiny/Components/Component.cs:6]: i was thinking of Moving IComponent - IspeakingComponent etc to VTTiny.Base
and Component and the implementations to VTTiny.DefaultImplementations so plugins don't have to reference the full solution/project and just the Base and/or Default

**Comment by naomiEve** [comment]
> @naomiEve can you commit your .editorconfig file with your code style, I'll just make a runner that auto formats all this on pull request.

Sorry, I actually don't have one (shamefully never knew it existed). I can try to make one and push it, though.

**Comment by naomiEve** [comment]
Sorry for leaving you hanging by the way, I'm having a very busy weekend. I'll try to sort out the editor config thing by the beginning of the next week.

**Comment by HugoVG** [comment]
Btw do you oppose having Dependancy injection/inversion or do you want everything just reference the class.

**Comment by HugoVG** [comment]
also i'll make a dotnet pullrequest formatter, but i'm afraid that without the editorconfig it'll change it to c# standard format

**Comment by naomiEve** [comment]
> also i'll make a dotnet pullrequest formatter, but i'm afraid that without the editorconfig it'll change it to c# standard format

This works by me for now.

**Comment by naomiEve** [comment]
> Btw do you oppose having Dependancy injection/inversion or do you want everything just reference the class.

Hmm, could you elaborate on this one a bit further? How would you add Dependency Injection to VTubeTiny?

**Comment by naomiEve** [comment]
https://user-images.githubusercontent.com/9293619/233849778-adf775d2-2867-4559-8033-e6cd9f5ce710.mp4
Unless I misunderstand something, the speaking is completely busted. I've set up a model as always, and my character does not change to any of the speaking textures, other than the one while blinking.

File: VTTiny/Plugins/PluginManager.cs
Comment (naomiEve) [inline VTTiny/Plugins/PluginManager.cs:38]: Why are we loading the plugin twice here?

File: VTTiny.Plugin.Discord/DiscordBot.cs
Comment (naomiEve) [inline VTTiny.Plugin.Discord/DiscordBot.cs:9]: This class needs to be in the VTTiny.Plugin.Discord namespace.

**Comment by HugoVG** [comment]
> https://user-images.githubusercontent.com/9293619/233849778-adf775d2-2867-4559-8033-e6cd9f5ce710.mp4 Unless I misunderstand something, the speaking is completely busted. I've set up a model as always, and my character does not change to any of the speaking textures, other than the one while blinking, instead staying only at the Idle state.

For some reason i'm not able to reproduce on pc 1 and on pc 2
did you load from a scene made before you pulled?
 

https://user-images.githubusercontent.com/64896329/233854318-409486e0-f9e3-496d-ab56-8cddee940e78.mp4

for me everything works without an issue on latest commit

File: VTTiny/Plugins/PluginManager.cs
Comment (HugoVG) [reply VTTiny/Plugins/PluginManager.cs:38]: still finding a good solution for this one, i'm debating between using Plugin framework, or a nuget package for loading it, right now (since it doesn't really do anything because there is no plugin's yet) it does nothing

File: VTTiny.Plugin.Discord/DiscordBot.cs
Comment (HugoVG) [reply VTTiny.Plugin.Discord/DiscordBot.cs:9]: in dotnet 6+ the main entrypoint is the Csproj name, but i'll explicitly put it there

**Comment by HugoVG** [comment]
Should i make a seperate issue for this?
it almost all ``RenderEditorGUI`` there is a memory with the ``ImGui.PushID($"{GetHashCode()}");`` lines, it is allocting a ton on the SOH, sometimes even targeting LOH if it has a lot of smaller components(audio, childeren, text change, multiple textures).

**Comment by naomiEve** [comment]
> > https://user-images.githubusercontent.com/9293619/233849778-adf775d2-2867-4559-8033-e6cd9f5ce710.mp4 Unless I misunderstand something, the speaking is completely busted. I've set up a model as always, and my character does not change to any of the speaking textures, other than the one while blinking, instead staying only at the Idle state.
> 
> For some reason i'm not able to reproduce on pc 1 and on pc 2 did you load from a scene made before you pulled?
> 2023-04-23.19-11-00.mp4
> 
> for me everything works without an issue on latest commit

Hmm, weird. Now it works fine for me.
Don't know what happened before.

**Comment by naomiEve** [comment]
> Should i make a seperate issue for this? it almost all `RenderEditorGUI` there is a memory with the `ImGui.PushID($"{GetHashCode()}");` (coming from `GetHashCode()`) lines, it is allocting a ton on the SOH, sometimes even targeting LOH if it has a lot of smaller components(audio, childeren, text change, multiple textures).
> 
> image
> 
> image
> 
> Changed this one already but i'll mention it anyway image

Yeah please do, the less allocations we do each frame, the better.

File: .idea/.idea.VTTiny/.idea/.gitignore
Comment (naomiEve) [inline .idea/.idea.VTTiny/.idea/.gitignore:1]: One minor nitpick, can we not merge this into the outermost .gitignore file? Do we need this folder to exist?

File: VTTiny.Plugin.Discord/Services/Token.cs
Comment (naomiEve) [inline VTTiny.Plugin.Discord/Services/Token.cs:9]: As before, please make properties PascalCased.

File: README.md
Comment (naomiEve) [inline README.md:51]: Nitpick but instead of leaving comments in the readme, please put these up as issues :)

**Review by naomiEve — COMMENTED** [review]
Minor nitpicks, the rest of the PR seems fine. There's some style things I have issues with, but we can iterate on those in-tree at this point.

Next time, please do split your PRs into smaller chunks though. Reviewing such a massive change in various parts of the projects is a bit hard ^^'

**Comment by naomiEve** [comment]
Also, I really like the separate viseme idea, I just wish we could define blinking sprites for each of the viseme.
This should come in a separate PR though.

File: .idea/.idea.VTTiny/.idea/.gitignore
Comment (HugoVG) [reply .idea/.idea.VTTiny/.idea/.gitignore:1]: It does need to exist and be committed for Rider, for some reason rider does not like normal gitignores

**Comment by HugoVG** [comment]
> Also, I really like the separate viseme idea, I just wish we could define blinking sprites for each of the viseme.
> This should come in a separate PR though.

I still have to find a way to not make it change visemes every frame, i tried, didn't work all that well, gonna give it a shot later.

**Comment by naomiEve** [comment]
I'll get it merged now, sorry for the delay.

Looking over the code, it seems fine. Any further nitpicks I have, or issues, I'll just fix in tree now. But, yeah, as before, please do try to make your PRs atomic next time. Reviewing this PR was a bit hard with all of the changes ^^'.

Thank you for your contribution, and welcome!

**Pull request merged**
